### ISSUE TITLE
Deploy Failure: Fetch Failed

### ISSUE THREAD

**Issue by JustinWestphal** [arunoda/meteor-up-legacy#700]
Another in the long string of deploy fail issues!
Here are my errors after everything else going smoothly:

Started TaskList: Deploy app 'MyApp' (linux)
[xxx] - Uploading bundle
[xxx] - Uploading bundle: SUCCESS
[xxx] - Setting up Environment Variables
[xxx] - Setting up Environment Variables: SUCCESS
[xxx] - Invoking deployment process
[xxx] x Invoking deployment process: FAILED

```
-----------------------------------STDERR-----------------------------------
tgz
npm ERR! fetch failed https://registry.npmjs.org/has-ansi/-/has-ansi-0.1.0.tgz
npm ERR! fetch failed https://registry.npmjs.org/ansi-regex/-/ansi-regex-0.2.1.tgz
npm ERR! network connect ETIMEDOUT
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network 
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! System Linux 3.13.0-48-generic
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "install"
npm ERR! cwd /opt/MyApp/tmp/bundle/programs/server
npm ERR! node -v v0.10.36
npm ERR! npm -v 1.4.28
npm ERR! syscall connect
npm ERR! code ETIMEDOUT
npm ERR! errno ETIMEDOUT
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /opt/MyApp/tmp/bundle/programs/server/npm-debug.log
npm ERR! not ok code 0
-----------------------------------STDOUT-----------------------------------
----------------------------------------------------------------------------
```

I can copy and paste those .tgz files into my browser and connect and download just fine, I can't see what the network problem would be. Tried this on a number of different networks, same issue each time. I'm not using a proxy.

Here's my mup.json:

``` javascript
{
  // Server authentication info
  "servers": [
    {
      "host": "xxx",
      "username": "ubuntu",
      // "password": "password"
      // or pem file (ssh based authentication)
      "pem": "/Users/justinwestphal/meteor/myappkeypair.pem"
    }
  ],

  // Install MongoDB in the server, does not destroy local MongoDB on future setup
  "setupMongo": true,

  // WARNING: Node.js is required! Only skip if you already have Node.js installed on server.
  "setupNode": true,

  // WARNING: If nodeVersion omitted will setup 0.10.36 by default. Do not use v, only version number.
  "nodeVersion": "0.10.36",

  // Install PhantomJS in the server
  "setupPhantom": false,

  // Show a progress bar during the upload of the bundle to the server.
  // Might cause an error in some rare cases if set to true, for instance in Shippable CI
  "enableUploadProgressBar": true,

  // Application name (No spaces)
  "appName": "MyApp",

  // Location of app (local directory)
  "app": "/Users/justinwestphal/meteor/MyApp",

  // Configure environment
  "env": {
    "ROOT_URL": "xxx",
    "PORT": 80,
    "MONGO_URL": "mongodb://localhost:27017/myapp-database"
  },

  // Meteor Up checks if the app comes online just after the deployment
  // before mup checks that, it will wait for no. of seconds configured below
  "deployCheckWaitTime": 15
}
```

**Comment by MasterJames** [comment]
Some have had success with clarifying to registry? Or not using the secure socket layer varient.
npm set registry http://registry.npmjs.org/
So with just http not httpS.
Or in your .npmrc file
registry=http://registry.npmjs.org

Also if there's any firewall you can disable or check your routing permissions with the provider's security network configuration.

Anyway it's a network problem as suggested in the error.

This might be of help
http://docs.aws.amazon.com/AmazonVPC/latest/GettingStartedGuide/getting-started-create-security-group.html

**Comment by JustinWestphal** [comment]
Great idea. I did this:

`sudo npm config set registry http://registry.npmjs.org/`

and running  `npm config get registry` yields:
http://registry.npmjs.org/

However, deploying still fails at the same step, 

``` javascript
    npm ERR! fetch failed https://registry.npmjs.org/ansi-styles/-/ansi-styles-1.1.0.tgz
    npm ERR! fetch failed https://registry.npmjs.org/asap/-/asap-2.0.3.tgz
```

Also, my security group has everything wiiiiiiide open. I'm not behind any sort of crazy firewall, but just to be certain, what ports need to be forwarded?
ideas?

**Comment by MasterJames** [comment]
Hmm obviously you set and get no SSL in the URL but it still shows httpS when you tried it? !
Maybe there's an override to check.
The four relevant files are:

per-project config file (/path/to/my/project/.npmrc)
per-user config file (~/.npmrc)
global config file ($PREFIX/etc/npmrc)
npm builtin config file (/path/to/npm/npmrc)
See npmrc(5) for more details.

The ports would be standard 80 and SSL 443 I recall.

It's worth a final check but I'm sure 'get' is get but maybe a project override exists?

You'll need to test the network conectivity from the target machine via SSH to the npm registry.

**Comment by JustinWestphal** [comment]
I can successfully SSH into my EC2 instance. Where are these .npmrc files? I have hidden files visible on my local machine and I've never seen those.

**Comment by MasterJames** [comment]
It's optional you can make a blank one if you want.
I'm certain they are searched in some order and they override each other. Basically your looking for the remote possibility there's a conflict hiding. 

Use a terminal and ls -lags .\* or whatever to reveal gui hidden files.

**Comment by JustinWestphal** [comment]
No good. I've put that .npmrc in every single directory listed above and various other for good measure. For the record, the contents are:

``` Text
registry=http://registry.npmjs.org/
strict-ssl=false
```

I still get the same error as before.
So to recap, I _can_ SSH successfully into my EC2 instance and I have set strict-ssl to false and changed the registry to http:// in the npm config.

There's nothing special about my app, and there's nothing odd or special about my EC2 account. I've never really delved into terminal or npm before this so there's no way I could have messed up any default settings. Has anybody successfully been able to use MUP to deploy a Meteor application to an Amazon EC2 instance on Mac OSX? If so, how did they circumvent this problem?

If there's no way around this, what are my other options for deploying my app onto my EC2 instance?

Edit: And now mup config is throwing the following error:

```
Meteor Up: Production Quality Meteor Deployments
------------------------------------------------

/usr/local/lib/node_modules/mup/bin/mup:28
    actionsRegistry[action]();
                           ^

TypeError: actionsRegistry[action] is not a function
    at runActions (/usr/local/lib/node_modules/mup/bin/mup:28:28)
    at Object.<anonymous> (/usr/local/lib/node_modules/mup/bin/mup:21:3)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:475:10)
    at startup (node.js:117:18)
    at node.js:951:3

```

**Comment by MasterJames** [comment]
Okay it wasn't that. The process of elimination is getting you closer to the answer. To bad it wasn't that.

The error clearly states a network connection timed out. If you can ssh into the target machine you should be able to wget or curl 
https://registry.npmjs.org/has-ansi/-/has-ansi-0.1.0.tgz
BTW Changing the httpS should not show up in the log then either.
If it can download it but not via mupx I'd be pretty surprised and equally confused.

Also looking at the settings maybe try changing the node version to 0.10.40 as well, this I think is more compatible with meteor 1.2.

Which ec2 instance type is it?
https://aws.amazon.com/ec2/instance-types/

**Comment by JustinWestphal** [comment]
Hi @MasterJames , sorry it's been a minute but I had to take a break from the project. @_@

I tried changing the node version to 0.10.40 as you said, but I get the same error. No good.

My instance type is     t2.micro.

I connected via SSH to my instance and initiated a curl command for the link you suggested, but was met with this error:

`curl: (7) Failed to connect to registry.npmjs.org port 443: Connection timed out`

Also, you keep mentioning mupx, emphasis on the 'x'. As far as I know, I just have mup. Is there another version I should be using?

Edit: Tried again opening the https 443 ports in my security group. Attached is a screenshot of the result.

![screen shot 2015-11-21 at 5 21 20 pm](https://cloud.githubusercontent.com/assets/2451149/11321443/63b3aec6-9074-11e5-8ff8-a174410c30da.png)

**Comment by MasterJames** [comment]
Try manually running the appropriate npm install command from the target.
Yes MupX  is here and the one I'm more familiar because it shifts to the newer concept of docker. Definitely try that.

https://github.com/arunoda/meteor-up/tree/mupx

Also look at quickly trying a larger then micro instance.
You can check disk space with df command too.

**Comment by JustinWestphal** [comment]
Okay, tried mupx instead. Here are a whole new batch of errors.

On `mupx init`:

```[ec2-52-27-114-154.us-west-2.compute.amazonaws.com] x Installing MongoDB: FAILED

``````
-----------------------------------STDERR-----------------------------------
Error response from daemon: no such id: mongodb
Error: failed to remove containers: [mongodb]
Error response from daemon: Cannot start container abef292579e3cd533f1d1ecc1c70630ec1038289a81c475391adbebfa900017b: failed to create endpoint mongodb on network bridge: Error starting userland proxy: listen tcp 127.0.0.1:27017: bind: address already in use
-----------------------------------STDOUT-----------------------------------
plete
d8263e0fcae7: Verifying Checksum
d8263e0fcae7: Download complete
d01299b0c7d3: Verifying Checksum
d01299b0c7d3: Download complete
1a64ad3ccff1: Verifying Checksum
1a64ad3ccff1: Download complete
68e42ff590bd: Verifying Checksum
68e42ff590bd: Download complete
68e42ff590bd: Pull complete
b4c4e8b590a7: Pull complete
f037c6d892c5: Pull complete
1a64ad3ccff1: Pull complete
7f85ac94fbfc: Pull complete
d01299b0c7d3: Pull complete
9a64a1ad8268: Pull complete
52d85577e44d: Pull complete
8f1cfd6db785: Pull complete
082b67a34646: Pull complete
0d4f8425d471: Verifying Checksum
0d4f8425d471: Download complete
0d4f8425d471: Pull complete
964b5afbd056: Pull complete
c976d54ea8ed: Pull complete
d8263e0fcae7: Pull complete
a08422dd6a11: Pull complete
99e2f2cde1c3: Pull complete
202e2c1fe066: Pull complete
Digest: sha256:223d59692269be18696be5c4f48e3d4117c7f11e175fe760f6b575387abc1bba
Status: Downloaded newer image for mongo:latest
abef292579e3cd533f1d1ecc1c70630ec1038289a81c475391adbebfa900017b
----------------------------------------------------------------------------```
``````

So I figured since I've run `mup setup` about a million times by now, there HAS to be a working mongo install on there somewhere, so I turned off the mongo install in mup.json for mupx. Then tried setup and succeeded. Here's what happened for deploy:

```[ec2-52-27-114-154.us-west-2.compute.amazonaws.com] x Installing MongoDB: FAILED

``````
-----------------------------------STDERR-----------------------------------
Error response from daemon: no such id: mongodb
Error: failed to remove containers: [mongodb]
Error response from daemon: Cannot start container abef292579e3cd533f1d1ecc1c70630ec1038289a81c475391adbebfa900017b: failed to create endpoint mongodb on network bridge: Error starting userland proxy: listen tcp 127.0.0.1:27017: bind: address already in use
-----------------------------------STDOUT-----------------------------------
plete
d8263e0fcae7: Verifying Checksum
d8263e0fcae7: Download complete
d01299b0c7d3: Verifying Checksum
d01299b0c7d3: Download complete
1a64ad3ccff1: Verifying Checksum
1a64ad3ccff1: Download complete
68e42ff590bd: Verifying Checksum
68e42ff590bd: Download complete
68e42ff590bd: Pull complete
b4c4e8b590a7: Pull complete
f037c6d892c5: Pull complete
1a64ad3ccff1: Pull complete
7f85ac94fbfc: Pull complete
d01299b0c7d3: Pull complete
9a64a1ad8268: Pull complete
52d85577e44d: Pull complete
8f1cfd6db785: Pull complete
082b67a34646: Pull complete
0d4f8425d471: Verifying Checksum
0d4f8425d471: Download complete
0d4f8425d471: Pull complete
964b5afbd056: Pull complete
c976d54ea8ed: Pull complete
d8263e0fcae7: Pull complete
a08422dd6a11: Pull complete
99e2f2cde1c3: Pull complete
202e2c1fe066: Pull complete
Digest: sha256:223d59692269be18696be5c4f48e3d4117c7f11e175fe760f6b575387abc1bba
Status: Downloaded newer image for mongo:latest
abef292579e3cd533f1d1ecc1c70630ec1038289a81c475391adbebfa900017b
----------------------------------------------------------------------------```
``````

EDIT: Tried setting Mongo Setup to false and leaving the mongo database url as "MONGO_URL": "mongodb://localhost:27017/appname-database"... here's the error after deploying:

x Verifying deployment: FAILED

```
-----------------------------------STDERR-----------------------------------
json meteor-dev-bundle@0.0.0 No README data
=> Starting meteor app on port:80

/bundle/bundle/programs/server/node_modules/fibers/future.js:278
                        throw(ex);
                              ^
Error: failed to connect to [localhost:27017]
npm WARN package.json meteor-dev-bundle@0.0.0 No description
npm WARN package.json meteor-dev-bundle@0.0.0 No repository field.
npm WARN package.json meteor-dev-bundle@0.0.0 No README data
=> Starting meteor app on port:80

/bundle/bundle/programs/server/node_modules/fibers/future.js:278
                        throw(ex);
                              ^
Error: failed to connect to [localhost:27017]
npm WARN package.json meteor-dev-bundle@0.0.0 No description
npm WARN package.json meteor-dev-bundle@0.0.0 No repository field.
npm WARN package.json meteor-dev-bundle@0.0.0 No README data
=> Starting meteor app on port:80

/bundle/bundle/programs/server/node_modules/fibers/future.js:278
                        throw(ex);
                              ^
Error: failed to connect to [localhost:27017]

=> Redeploying previous version of the app

-----------------------------------STDOUT-----------------------------------

To see more logs type 'mup logs --tail=50'

----------------------------------------------------------------------------
```

Followup question: I've invoked deploy and failed at least several dozen times with mup and mupx. It's reuploading the whole package each time, right? Can I delete all the extraneous copies somehow to free up disk space?

**Comment by MasterJames** [comment]
Did you follow the migration guide?

**Comment by MasterJames** [comment]
Yes you should purge /tmp and docker logs can fill up too.

**Comment by JustinWestphal** [comment]
When I SSH in, in the ubuntu directory there are no other files or folders where I suppose meteor is supposed to be, so none of the commands in the migration guide were applicable.

**Comment by MasterJames** [comment]
The text before your curl in the screen grab was looking pretty weird? ! Is that the prompt?  Some other language? 
Anyway are you using Ubuntu 14 for the target system?

**Comment by JustinWestphal** [comment]
That was the raw text I got. Yeah just jumbled characters. This is all out-of-the-box, too; I've never used AWS before following the tutorials to get it set up for Meteor. Here's my Ubuntu version.

Ubuntu 14.04.2 LTS (GNU/Linux 3.13.0-48-generic x86_64)

**Comment by JustinWestphal** [comment]
Holy crap I finally figured it out.

I went back to solve the earlier problem of being unable to install mongodb. The error was:
`bind: address already in use`
I solved it by SSHing in and running this command:
`ps wuax | grep mongo`
then doing a `sudo kill XXXX` for the ID of the mongodb instance that was running.

I then ran `mupx setup` and `mupx deploy` and they both executed flawlessly.

Now the app is up and I can access it just fine! Thank you so much for all your help, @MasterJames! I hope these solutions help others down the line.

**Comment by justindra** [comment]
@JustinWestphal So just wanted to confirm you were never able to do it with `mup`?

**Comment by sahanDissanayake** [comment]
Yes, Can deploy with mup, but not with mupx

**Comment by waynem59** [comment]
@JustinWestphal 23 Nov 2015: YOU SAVED MY A**E .. thanks a lot. But I got an additional error: 

> "Failed to remove container (meteorapp-frontend): Error response from daemon: No such container: meteorapp-frontend docker: Error response from daemon: failed to create endpoint meteorapp on network bridge: Error starting userland proxy: listen tcp 0.0.0.0:80: bind: address already in use.". 

After opening the door by killing the mongod process the solution was finally: 
- Delete the content of /tmp via SSH
- Stopping the meteorapp process (meteorapp is my Meteor app ;-))
- Removing /opt/meteorapp as sudo (sudo rm -R -f /opt/meteorapp)
- mupx setup on my local machine
- mupx deploy

Now everything seemed to be ok. On the first look. Opening the app in the browser I got a blank screen. The reason was that the node version was not up to date. So I had to update it manually. Again via SSH: 
- sudo npm cache clean -f
- sudo npm install -g n
- sudo n stable

Now everything works fine.
